sdk/highlight-next/src/util/with-highlight-nodejs-app-router.ts

```diff
@@ -1,27 +1,56 @@
 import type { NextRequest } from 'next/server'
 
-import { H, NodeOptions } from '@highlight-run/node'
+import { H, parseHeaders, NodeOptions } from '@highlight-run/node'
+import { IncomingHttpHeaders } from 'http'
 
 type NextContext = { params: Record<string, string> }
 type NextHandler<Body = unknown> = (
 	request: NextRequest,
 	context: NextContext,
+	highlight: HighlightInitReturnType,
 ) => Promise<Response>
+type HighlightInitReturnType = ReturnType<typeof H.init>
 
 export function Highlight(options: NodeOptions) {
-	H.init(options)
+	const NodeH = H.init(options)
+
 	return (originalHandler: NextHandler) =>
 		async (request: NextRequest, context: NextContext) => {
+			if (!NodeH) throw new Error('Highlight not initialized')
+
+			const { secureSessionId, requestId } = NodeH?.setHeaders(
+				request.headers,
+			)
+
 			try {
-				// Must await originalHandler to catch the error at this level
-				return await H.runWithHeaders(request.headers, async () => {
-					return await originalHandler(request, context)
-				})
-			} catch (error) {
-				const { secureSessionId, requestId } = H.parseHeaders(
+				return await H.runWithHeaders<Promise<Response>>(
 					request.headers,
-				)
+					async () => {
+						return new Promise<Response>((resolve, reject) => {
+							NodeH?.tracer.startActiveSpan(
+								'with-highlight-nodejs-app-router',
+								async (span) => {
+									try {
+										const result = await originalHandler(
+											request,
+											context,
+											NodeH,
+										)
 
+										span?.end()
+
+										await NodeH?.flush()
+
+										return resolve(result)
+									} catch (error) {
+										reject(error)
+									}
+								},
+							)
+						})
+					},
+				)
```

Comment:
as discussed offline, let's switch this to the approach of populating the trace id -> highlight ctx map and applying the highlight ctx on end of spans.